pulsar/client/client.py

```diff
@@ -999,6 +1004,130 @@ def _raw_check_complete(self):
+        return gcp_job_params
+
+
+class GcpMessageCoexecutionJobClient(BaseMessageCoexecutionJobClient, LaunchesGcpContainersMixin):
+    """A client that co-executes pods via GCP and depends on amqp for status updates."""
+
+    def __init__(self, destination_params, job_id, client_manager):
+        super().__init__(destination_params, job_id, client_manager)
+
+
+class GcpPollingCoexecutionJobClient(BasePollingCoexecutionJobClient, LaunchesGcpContainersMixin):
+    """A client that co-executes pods via GCP and doesn't depend on amqp."""
+
+    def __init__(self, destination_params, job_id, client_manager):
+        super().__init__(destination_params, job_id, client_manager)
+
+    def kill(self):
+        gcp_job_params = self._gcp_job_params
+        delete_gcp_job(gcp_job_params.project_id, gcp_job_params.region, self._tes_job_name, gcp_job_params.credentials_file)
+
+    def clean(self):
+        pass
+
+    def raw_check_complete(self) -> Dict[str, Any]:
+        gcp_job_params = self._gcp_job_params
+        job = get_gcp_job(gcp_job_params.project_id, gcp_job_params.region, self._tes_job_name, gcp_job_params.credentials_file)
+        status = job.status
+        state = status.state
+        return {
+            "status": gcp_state_to_pulsar_status(state),
+            "complete": "true" if gcp_state_is_complete(state) else "false",  # Ancient John, what were you thinking?
+        }
+
+
+def gcp_state_to_pulsar_status(state: Optional["batch_v1.JobStatus.State"]) -> str:
+    state = state or batch_v1.JobStatus.State.STATE_UNSPECIFIED
+    # STATE_UNSPECIFIED	Job state unspecified.
+    # QUEUED	Job is admitted (validated and persisted) and waiting for resources.
+    # SCHEDULED	Job is scheduled to run as soon as resource allocation is ready. The resource
+    #           allocation may happen at a later time but with a high chance to succeed.
+    # RUNNING	Resource allocation has been successful. At least one Task in the Job is RUNNING.
+    # SUCCEEDED	All Tasks in the Job have finished successfully.
+    # FAILED	At least one Task in the Job has failed.
+    # DELETION_IN_PROGRESS	The Job will be deleted, but has not been deleted yet. Typically this is because resources
+    #                       used by the Job are still being cleaned up.
+    # CANCELLATION_IN_PROGRESS	The Job cancellation is in progress, this is because the resources used by the Job are
+    #                           still being cleaned up.
+    # CANCELLED The Job has been cancelled, the task executions were stopped and the resources were cleaned up.
+    state_map = {
+        batch_v1.JobStatus.State.STATE_UNSPECIFIED: manager_status.FAILED,
+        batch_v1.JobStatus.State.QUEUED: manager_status.PREPROCESSING,
+        batch_v1.JobStatus.State.SCHEDULED: manager_status.PREPROCESSING,
+        batch_v1.JobStatus.State.RUNNING: manager_status.RUNNING,
+        batch_v1.JobStatus.State.SCHEDULED: manager_status.COMPLETE,
```

Comment:
The key `batch_v1.JobStatus.State.SCHEDULED` appears twice in the `state_map` (and in the `state_map` in `gcp_state_is_complete` below.

Reply:
Good catch - thank you!